mitiq/rem/rem.py

```diff
@@ -0,0 +1,173 @@
+# Copyright (C) 2021 Unitary Fund
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+"""Readout Confusion Inversion."""
+
+from typing import Callable, Iterable, List, Optional, Sequence, Tuple, Union
+
+from functools import partial, wraps
+import numpy as np
+
+from cirq.sim import sample_state_vector as sample_prob_dist
+from cirq.qis.states import to_valid_state_vector as to_prob_dist
+
+from mitiq._typing import QPROGRAM, MeasurementResult
+from mitiq.executor.executor import Executor
+from mitiq.observable.observable import Observable
+
+MatrixLike = Union[
+    np.ndarray,
+    Iterable[np.ndarray],
+    List[np.ndarray],
+    Sequence[np.ndarray],
+    Tuple[np.ndarray],
+]
+
+
+def mitigate_measurements(
+    noisy_result: MeasurementResult,
+    inverse_confusion_matrix: MatrixLike,
+) -> MeasurementResult:
+    """Applies the inverse confusion matrix against the noisy measurement
+    result and returns the adjusted measurements.
+
+    Args:
+        noisy_results: The unmitigated ``MeasurementResult``.
+        inverse_confusion_matrix: The inverse confusion matrix to apply to the
+            probability vector estimated with noisy measurement results.
+    """
+    assert noisy_result.qubit_indices is not None
```

Comment:
In addition to this assertion, we could raise errors if the size of the inverse confusion matrix is not compatible with the measurement results. E.g. different number of qubits...

Reply:
I've added an additional assert and a new test.